chronos/asyncloop.nim

```diff
@@ -256,12 +263,47 @@ template processTimersGetTimeout(loop, timeout: untyped) =
     if len(loop.callbacks) != 0:
       timeout = 0
 
-declareCounter chronos_loop_timers, "loop timers"
-declareGauge chronos_loop_timers_queue, "loop timers queue"
+when defined(metrics):
+  declareCounter chronos_loop_timers, "loop timers"
+  declareGauge chronos_loop_timers_queue, "loop timers queue"
+  declareCounter chronos_poll_ticks, "Chronos event loop ticks"
+  declareCounter chronos_poll_events, "Chronos poll events", ["event"]
+  declareCounter chronos_future_callbacks, "Future callbacks", ["location"]
+  declareCounter chronos_processed_callbacks, "total number of processed callbacks"
+
+
+  proc processCallbacksByFuture() =
```

Comment:
What the reason to count callbacks per Future?
Almost every possible Future[T] will have only 1 or 2 callbacks, in rare cases it will have more then 2 (allFutures, one), but i really do not understand why this code ever needed?

Reply:
> What the reason to count callbacks per Future?

Per Future type, not instance.

> but i really do not understand why this code ever needed

So I can see which Future types are used more often. In NBC's case, it's those produced by readExactly() and write(), peaking at 140-180 per second, every few seconds.

It shows inefficient internal reads and writes, because they correspond to peaks of 20 OS read events and 50 write events per second.